hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java

```diff
@@ -588,10 +591,21 @@ private void deleteCacheResources(String name) {
     }
 
     @Override
+    public Operation getPreJoinOperation() {
+        Version clusterVersion = nodeEngine.getClusterService().getClusterVersion();
+        OnJoinCacheOperation preJoinCacheOperation = null;
+        if (clusterVersion.isGreaterOrEqual(V3_9)) {
```

Comment:
Before join, `ClusterService.getClusterVersion()` returns `Version.UNKNOWN`. Pre- & post-join ops are gathered in the context of:
- responding to a [`WhoIsMasterOp`](https://github.com/vbekiaris/hazelcast/blob/b1a61da0701279c7db39f6df8666112a4e0c36ef/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterJoinManager.java#L644-L644)
- processing a [`JoinRequest`](https://github.com/vbekiaris/hazelcast/blob/b1a61da0701279c7db39f6df8666112a4e0c36ef/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterJoinManager.java#L239-L239)

In both cases, my understanding is that the member preparing the pre-/post-join operations should be the master so cluster version will have been already set.